### PR TITLE
Feature: linkcheck-action

### PR DESCRIPTION
**Pull request by HunterGerlach**
- Add support for local build and linkchecking of docs
- Add GitHub Action to automatically check links before prior to publication
- Minor reformat of Contributing readme
- Fix existing broken links

### CODE REVIEW

File: .github/workflows/linkcheck.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+env:
+  POETRY_VERSION: "1.3.1"
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version:
```
Comment (hwchase17) [inline .github/workflows/linkcheck.yml:16]: probably only needs to run on one python version

**Review by hwchase17 — COMMENTED** [review]
yes! love this - thank you so much!!!!

File: .github/workflows/linkcheck.yml (hunk 1 above)
Comment (HunterGerlach) [reply .github/workflows/linkcheck.yml:16]: Good call.

**Comment by HunterGerlach** [comment]
@hwchase17 I am running into an issue with the link re-writes. Since they are rewritten from `.rst`, `.md`, and `.ipynb`, they are unable to be found by `linkcheck`. I put a brute force example in the `conf.py` of how we _could_ solve this if necessary, but I would much rather have a clean method. Unfortunately, I'm not too well-versed on Sphinx to know if there's a "right" way to do this or not. Any ideas you have would be appreciated.

**Comment by HunterGerlach** [comment]
I don't mind doing it to free you up for other more important tasks if you just want to do a few so I can see an example or two (e.g. `.md` and `.ipynb`).

**Comment by hwchase17** [comment]
> I don't mind doing it to free you up for other more important tasks if you just want to do a few so I can see an example or two (e.g. `.md` and `.ipynb`).

i *think* (not 100%, need to spend some time to be sure) that ":doc:`foo <utils/how_to_guides>`:" should work?

**Comment by HunterGerlach** [comment]
We'll see...

**Comment by HunterGerlach** [comment]
> > I don't mind doing it to free you up for other more important tasks if you just want to do a few so I can see an example or two (e.g. `.md` and `.ipynb`).
> 
> i _think_ (not 100%, need to spend some time to be sure) that ":doc:`foo <utils/how_to_guides>`:" should work?

I was unable to get it to work with `make linkcheck` - I think it's because it's checking the pre-compiled links _before_ their extensions change and it gets confused.

I switched to using a dedicated linkchecker instead.

I have removed all broken links, so running this from now on should prevent any broken links from getting into `master`.

**Comment by hwchase17** [comment]
@HunterGerlach .... i dont think it is an overstatement to call you a hero for this

im gonna merge this into a dev branch and build a copy of the docs from that (just to check that the build-the-docs action works as expected with this)

**Pull request merged**